mobile/examples/cc/fetch_client/fetch_client_main.cc

```diff
@@ -1,22 +1,112 @@
+    std::cerr << response_body << "\n";
+    if (end_stream) {
+      std::cerr << "Received final data\n";
+    }
+  };
+  stream_callbacks.on_complete_ =
+      [request_finished, used_protocols](envoy_stream_intel, envoy_final_stream_intel final_intel) {
+        std::cerr << "Request finished after "
+                  << final_intel.stream_end_ms - final_intel.stream_start_ms << "ms\n";
+        used_protocols->push_back(static_cast<Envoy::Http::Protocol>(final_intel.upstream_protocol));
+        request_finished->Notify();
+      };
+  stream_callbacks.on_error_ = [request_finished, status](const Envoy::EnvoyError& error,
+                                                            envoy_stream_intel,
+                                                            envoy_final_stream_intel final_intel) {
+    *status = ENVOY_FAILURE;
+    std::cerr << "Request failed after " << final_intel.stream_end_ms - final_intel.stream_start_ms
+              << "ms with error message: " << error.message_ << "\n";
+    request_finished->Notify();
+  };
+  stream_callbacks.on_cancel_ = [request_finished](envoy_stream_intel,
+                                                    envoy_final_stream_intel final_intel) {
+    std::cerr << "Request cancelled after "
+              << final_intel.stream_end_ms - final_intel.stream_start_ms << "ms\n";
+    request_finished->Notify();
+  };
+  return stream_callbacks;
+}
+
 // Fetches each URL specified on the command line in series,
 // and prints the contents to standard out.
 int main(int argc, char** argv) {
-  Envoy::Fetch client;
   std::vector<absl::string_view> urls;
   // Start at 1 to skip the command name.
   for (int i = 1; i < argc; ++i) {
     urls.push_back(argv[i]);
   }
-  std::vector<Envoy::Http::Protocol> protocols;
-  client.fetch(urls, /* quic_hints=*/{}, /* protocols= */ protocols);
+
+  // Build and run engine.
+  absl::Notification engine_running;
+  Envoy::Platform::EngineSharedPtr engine = Envoy::Platform::EngineBuilder()
+      .setLogLevel(Envoy::Logger::Logger::trace)
+      .addRuntimeGuard("dns_cache_set_ip_version_to_remove", true)
+      .addRuntimeGuard("quic_no_tcp_delay", true)
+      .setOnEngineRunning([&engine_running]() { engine_running.Notify(); })
+      .build();
+  engine_running.WaitForNotification();
+
+  // Iterate over the input URLs, reusing the engine.
+  for (const auto url_string : urls) {
+    Envoy::Http::Utility::Url url;
+    if (!url.initialize(url_string, /*is_connect_request=*/false)) {
+      std::cerr << "Unable to parse url: '" << url_string << "'\n";
+      return ENVOY_FAILURE;
+    }
+    std::cout << "Fetching url: " << url.toString() << "\n";
+
+    // Craete stream.
+    envoy_status_t status = ENVOY_SUCCESS;
+    absl::Notification request_finished;
+    std::vector<Envoy::Http::Protocol> used_protocols;
+    Envoy::Platform::StreamSharedPtr stream = engine->streamClient()->newStreamPrototype()->start(
```

Comment:
Based on the original example, it seems like `StreamPrototype::start()` may not be thread safe? In this case I'm running them sequentially, so I think there's no issue, but I could add a comment.